### DISCUSSION TITLE
An error occurred during send TCP : Other Socket error (32)'

### DISCUSSION THREAD

**Discussion by martin2day2** [gijzelaerr/python-snap7#480] [General]
Hello,

i use a Pi with python to send data from my heatpump to my S7-300.
The program works good but after some hours of working i get an error:

```
Dec 10 14:34:09 raspberrypi python3[12962]: fehler
Dec 10 14:34:14 raspberrypi python3[12962]: b' ISO : An error occurred during send TCP : Other Socket error (32)'
```
this is how my program is looks like.
```
import snap7
...
self.plc = snap7.client.Client()
self.plc.connect("192.168.10.100", 0, 2)
...
def heater(self)
            wdata = bytearray(4)
            util.set_int(wdata, 0, self.grid_p)
            util.set_int(wdata, 2, int(self.actTemp_water * 10))
            try:
                self.plc.db_write(db_number=201, start=0, data=wdata)
            except:
                print('fehler')
            time.sleep(5)

```

Please help me to get my code work.
What does the fault mean?

Have a nice day
Martin

**Comment by spreeker** [comment]
The try - except now does not print the pull stack trace and error. 

use logging.exception like here: https://stackoverflow.com/questions/4508849/how-to-log-python-exception

```
import logging 
def foo():
    try:
        some_code()
    except:
        logging.exception('')
```

with out more information it is hard to debug? how often do you send water temperature?

**Reply by martin2day2** [reply]
Hello,

okay i change the code.
Today i get enother error:

```
Dec 12 15:26:43 raspberrypi python3[26512]:     self.run()
Dec 12 15:26:43 raspberrypi python3[26512]:   File "/usr/lib/python3.9/threading.py", line 892, in run
Dec 12 15:26:43 raspberrypi python3[26512]:     self._target(*self._args, **self._kwargs)
Dec 12 15:26:43 raspberrypi python3[26512]:   File "/home/martin/app.py", line 61, in heater
Dec 12 15:26:43 raspberrypi python3[26512]:     self.plc.db_write(db_number=201, start=0, data=self.wdata)
Dec 12 15:26:43 raspberrypi python3[26512]:   File "/home/martin/.local/lib/python3.9/site-packages/snap7/client/__init__.py", line 24, in f
Dec 12 15:26:43 raspberrypi python3[26512]:     check_error(code, context="client")
Dec 12 15:26:43 raspberrypi python3[26512]:   File "/home/martin/.local/lib/python3.9/site-packages/snap7/common.py", line 89, in check_error
Dec 12 15:26:43 raspberrypi python3[26512]:     raise RuntimeError(error)
Dec 12 15:26:43 raspberrypi python3[26512]: RuntimeError: b' ISO : An error occurred during recv TCP : Connection timed out'
```

i have allready an try befor the self.plc.db_write(db_number=201, start=0, data=self.wdata)...
What can i do here?

Martin

**Comment by spreeker** [comment]
Hi,

I see you use threading which complicates things. The connection is lost
which can happen on a network / wifi connection.

Some cleanup must probably be done before (re) connection. On network
failure sleep a few seconds and try again later.

- Probably A new connection must be build to the PLC on connection loss.

I do not know enough of you code and context.

Every x time you do an update also a new connexion is needed. Now I assume
one connection is made and used for ours and is lost. If you make and close
a connection each time this is probably more reliable.

Cheers,

Stephan.

**Comment by gijzelaerr** [comment]
Closing as stale. Socket error 32 (broken pipe) indicates the TCP connection was lost. Reconnect on connection loss.
